### PR TITLE
Serve deployment dashboard

### PR DESCRIPTION
**Pull request by alanwguo**
## Why are these changes needed?
1. Adds a serve_deployment dashboard exported by ray
2. Adds links to the serve dashboard and the serve deployment dashboard in the Serve section of the Ray Dashboard if grafana is running
3. Add embedded graphs to the Serve page and the Serve Replicas page if grafana is running

Deployment dashboard
Screenshot 2023-03-22 at 6 37 56 PM

Grafana is running:
Screenshot 2023-03-23 at 8 24 42 PM
Screenshot 2023-03-23 at 8 25 27 PM
Screenshot 2023-03-23 at 8 24 37 PM

Grafana not running:
Screenshot 2023-03-23 at 8 25 05 PM
Screenshot 2023-03-23 at 8 25 19 PM
Screenshot 2023-03-23 at 8 25 12 PM

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
I did not review the code, leaving that to others, but the dashboard/plots LGTM.

**Comment by scottsun94** [comment]
The dashboards/plots LGTM!

Just OOC, have one question about the graph content:
Why are there two similar routes `foo/ `and `foo` in the `QPS per route` graph in serve page?

**Comment by alanwguo** [comment]
> The dashboards/plots LGTM!
> 
> Just OOC, have one question about the graph content: Why are there two similar routes `foo/ `and `foo` in the `QPS per route` graph in serve page?

@Sihan and I are syncing about this today. It looks like a double counting

**Comment by akshay-anyscale** [comment]
Plots look good. One question - are the deployment metrics (across replicas) visible on any page? I only see route level and replica level metrics in the plots

**Comment by alanwguo** [comment]
> Plots look good. One question - are the deployment metrics (across replicas) visible on any page? I only see route level and replica level metrics in the plots

Not for ray 2.4. We can add it as we get feedback from users.

File: dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import { GlobalContext } from "../../App";
+import { CollapsibleSection } from "../../common/CollapsibleSection";
+import { ClassNameProps } from "../../common/props";
+import {
+  MetricConfig,
+  TIME_RANGE_TO_FROM_VALUE,
+  TimeRangeOptions,
+} from "../metrics";
+
+const useStyles = makeStyles((theme) =>
```
Comment (galenhwang) [inline dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx:20]: nit: extract common serve/serve deployment metrics section parts

File: dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  const toParam = to !== null ? `&to=${to}` : "";
+  const timeRangeParams = `${fromParam}${toParam}`;
+
+  const replicaButtonUrl = useViewServeDeploymentMetricsButtonUrl(
+    deploymentName,
+    replicaId,
+  );
+
+  return grafanaHost === undefined ||
+    !prometheusHealth ||
```
Comment (galenhwang) [inline dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx:116]: nit: these two can be removed, they're checked in `useViewServeDeploymentMetricsButtonUrl()`

File: dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+      flexWrap: "wrap",
+      gap: theme.spacing(3),
+      marginTop: theme.spacing(2),
+    },
+    chart: {
+      width: "100%",
+      height: 400,
+      overflow: "hidden",
+      [theme.breakpoints.up("md")]: {
+        // Calculate max width based on 1/3 of the total width minus padding between cards
```
Comment (galenhwang) [inline dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx:35]: nit: `gap` between cards

**Review by galenhwang — APPROVED** [review]
LGTM

**Comment by alanwguo** [comment]
> The dashboards/plots LGTM!
> 
> Just OOC, have one question about the graph content: Why are there two similar routes `foo/ `and `foo` in the `QPS per route` graph in serve page?

Finished pairing with @sihanwang41 , details are in this ticket: https://github.com/ray-project/ray/issues/33686
TL;DR: Don't think we need to address anything for ray 2.4

File: dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx (hunk 2 above)
Comment (alanwguo) [reply dashboard/client/src/pages/serve/ServeDeploymentMetricsSection.tsx:116]: The check is necessary for typescript type inference to work. Otherwise it thinks these values are possibly undefined.

**Review by sihanwang41 — APPROVED** [review]
Graph lgtm!

**Comment by rkooo567** [comment]
Lmk when it is ready to merge. I will rely on galen & sihan for the PR review

**Comment by alanwguo** [comment]
It's ready. I think master is pretty red right now. All failing tests are unrelated
